### ISSUE TITLE
Warn when operators are used without parens after infix function

### ISSUE THREAD

**Issue by schoettl** [ndmitchell/hlint#173]
Example:

``` haskell
func f a b = a `f` b/2
```

The missing space around `/` suggests that `b` is divided by `2` _before_ function application.

But function application (here infix) has still highest precedence and so the example is equivalent to:

``` haskell
func f a b = (f a b) / 2
```

Can HLint produce a warning in such a case, especially when there is no space around the operator (`/`)?

It take me long to find this bug in my code and I am sure other programmers made the same type of bug before. So it would be a nice feature for HLint.

**Comment by ndmitchell** [comment]
I like the idea of hinting when the visual operator precedence and the real operator precedence don't match. Unfortunately, I'm not sure if it's feasible for HLint:

* You can give fixities to functions, so if you give `f` the right fixity, it actually is grouped correctly: http://stackoverflow.com/questions/8139066/haskell-infix-function-application-precedence
* HLint doesn't track imports from different modules, so if `f` is imported, HLint won't know what fixity it has.

There are a few cases I can still catch, specifically:

* Using only operators from base, e.g. `a * c+1`
* Using function application, e.g. `f x+1`

So given I can't tackle the full generality, the question is if the things I can catch makes it worth doing or not, since it's a fairly tricky change to program up, and might not catch much.

**Comment by schoettl** [comment]
OK I understand. It's tricky to make it general. But I am happy to hear that spaces can be considered by HLint! Good work :)

How about a warning like this:
```
Make sure that the infix function "f" has lower precedence then "b/2"  
or use parenthesis around "b/2". Default presedence for infix functions  
is very high if not changed by infixl or infixr.
```

I knew about fixities for functions. But I think infix functions that are not from base (like `mod` and `div`) could be simply treated as default infix functions without changed presedence.

Because my understanding is -- unless I write a framework -- it is considered as bad practice to define own operators and change default precedence of functions.

> Defining symbolic infix identifiers should be left to library writers only.

-- https://wiki.haskell.org/Programming_guidelines

Anyway, I would really appreciate warnings for your suggestet cases (`a * b+1` and `f x+1`)

**Comment by ndmitchell** [comment]
I don't want a warning that can't be solved by changing your code, and things like "check the fixity" aren't helpful - since they might be wrong people ignore them. Doing it only for things like `mod`/`div` might be a better option there, since then I can always give a good warning.

I disagree with the infix identifier rule, partly because I don't see any difference between writing a library and a program (for example, HLint is a program, which is structured as about 6 API's internally, and then exposes a further API). I think a better rule is "be careful, try not to capture the spirit of Perl in your Haskell".

**Comment by schoettl** [comment]
You're right, there should not be warnings that cannot be resolved.

BTW are warnings for cases with `mod` general enough? Because one could also import another `mod` funciton with different presedence.

**Comment by ndmitchell** [comment]
You could import another `mod`, but if you do then you end up confusing yourself as well, since it's available in the Prelude by default. I'm happy to be wrong in relatively obscure corner cases. I also have some logic for checking if `mod` seems reasonably likely to have come from `Prelude`.

**Comment by ndmitchell** [comment]
The benefit seems not worth the cost.
